File: algo/CountingSort/countingSort_reverse.py
class countingSort:
    @classmethod
    def countingSort_reverse(cls, arr):
        if arr is None or len(arr) <= 1:
            return arr
        max = arr[0]
        min = arr[0]
        for i in range(1, len(arr)):
            if arr[i] > max:
                max = arr[i]
            elif arr[i] < min:
                min = arr[i]
        total_range = max - min + 1
        counting = [0] * total_range
        for element in arr:
            idx = element - min #why min?? test arr by hand will know
            counting[idx] += 1
        print(counting)
        counting[0] -= 1 #why?
        for i in range(1, total_range):
            counting[i] += counting[i-1]
        print(counting)
        result = [0] * len(arr)
        for i in range(len(arr)-1, -1, -1):
            position = counting[arr[i] - min]
            result[position] = arr[i]
            counting[arr[i] - min] -= 1
        for i in range(len(arr)):
            arr[i] = result[i]
if __name__ == "__main__":
    arr = [8, 7, 2, 8, 6, 8, 2]
    arr2 = [7, 8, 9, 7, 6, 7, 6, 8, 6, 6]
    countingSort.countingSort_reverse(arr)
    print(arr)
"""
[0, 0, 0, 0, 0, 0, 0]
[2, 0, 0, 0, 1, 1, 3]
[0, 2, 2, 2, 2, 3, 4]
 ↑
val = 2
"""
"""
代碼出處:
https://leetcode.cn/leetbook/read/sort-algorithms/ozyo63/
"""

File: algo/CountingSort/CountingSort_reverse.java
package CountingSort;

import java.util.Arrays;
public class CountingSort_reverse {
    void CountingSort(int[] arr){
        int min = arr[0];
        int max = arr[0];
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] < min)
                min = arr[i];
            else if (arr[i] > max)
                max = arr[i];
        }

        final int true_range = max - min + 1;
        int[] counting = new int[true_range]; //注意, 要記得counting的大小是用true_range

        for (int element : arr) {
            counting[element - min] ++;
        }
        System.out.println(Arrays.toString(counting));

        for (int index = 1; index < true_range; index++)//注意是true_range, 而不是counting.length
            counting[index] += counting[index - 1];
        for (int index = 0; index < true_range; index++)
            counting[index]--;
        System.out.println(Arrays.toString(counting));

        //寫第二次還是寫不出來
        int[] result = new int[arr.length];
        for (int index = arr.length - 1; index >= 0; index--){
            int val = arr[index];
            result[counting[val - min]] = val; //val - min 是在 counting 中的偏移位置, 可以從上兩個迴圈推得
            counting[val - min]--;
        }
        System.arraycopy(result, 0, arr, 0, result.length);
    }
    public static void main(String[] args) {
        int[] arr = {2, 4, 5, 2, 6, 4};
        CountingSort_reverse Sort = new CountingSort_reverse();
        Sort.CountingSort(arr);
        System.out.println(Arrays.toString(arr));
    }
}
